src/indexing.jl

```diff
@@ -488,6 +483,80 @@ end
     end
 end
 
+
+_getindex_kwargs(x, kwargs, args...) = @inbounds getindex(x, args...; kwargs...)
+
+function _generate_unsafe_getindex!_body(N::Int)
+    quote
+        Base.@_inline_meta
+        D = eachindex(dest)
+        Dy = iterate(D)
+        @inbounds Base.Cartesian.@nloops $N j d->I[d] begin
+            # This condition is never hit, but at the moment
+            # the optimizer is not clever enough to split the union without it
+            Dy === nothing && return dest
+            (idx, state) = Dy
+            dest[idx] = Base.Cartesian.@ncall $N _getindex_kwargs src kwargs j
+            Dy = iterate(D, state)
+        end
+        return dest
+    end
+end
+
+@generated function _unsafe_getindex!(dest::AbstractArray, src::AbstractArray, I::Vararg{Union{Real, AbstractArray}, N}; kwargs...) where N
+    _generate_unsafe_getindex!_body(N)
+end
+
+function unsafe_get_collection(::CPUIndex, src, inds; kwargs...)
+    return unsafe_get_collection_by_index(IndexStyle(src), src, inds)
+end
+
+# TODO
+function unsafe_get_collection_by_index(::IndexLinear, src, inds; kwargs...)
+    dst = similar(src, Base.index_shape(inds...))
+    src_iter = @inbounds(view(LinearIndices(src), inds...)) # FIXME not ideal iterator
+    dst_iter = indices(dst)
+    # TODO should we check that src_iter and dst_iter have the same size?
+    # at this point in the pipeline it shouldn't be an issue
+    src_i = iterate(src_itr)
+    dst_i = iterate(dst_itr)
+    @inbounds while src_i !== nothing
+        dst[dst_i] = src[src_i]
+        src_i = iterate(src, last(src_i))
+        dst_i = iterate(dst, last(dst_i))
+    end
+    return dst
+end
+
+function unsafe_get_collection_by_index(::IndexStyle, src, inds; kwargs...)
+    axs = to_axes(src, inds)
+    dest = similar(src, axs)
+    if map(Base.unsafe_length, axes(dest)) == map(Base.unsafe_length, axs)
+        # usually a generated function, don't allow it to impact inference result
+        _unsafe_getindex!(dest, src, inds...; kwargs...)
+    else
+        Base.throw_checksize_error(dest, axs)
+    end
+    return dest
+end
+
+function unsafe_get_collection(::CPUPointer, src, inds; kwargs...)
+    sz = static_length.(inds)
+    return unsafe_get_collection_by_pointer(          # FIXME not an actual function
+        allocate_memory(src, sz),                     # FIXME not an actual function
+        OptionallyStaticUnitRange(One(), prod(sz)),
+        pointer(src),
+        @inbounds(view(LinearIndices(src), inds...))  # FIXME not ideal iterator
+    )
+end
```

Comment:
Should we simply call `unsae_get_collection_by_pointer(pointer(src), src, inds)` here? I'd like to get as much of the indexing defined generically here as possible, and then maybe LoopVectorization could specialize based on pointer type. It would be nice to atleast get this working with `Array`, but perhaps optimizing that is too much for this package?

Reply:
You mean should the yet too be implemented `unsafe_get_collection_by_pointer` function just take those as arguments? That's fair, it could call `static_length` and allocate memory itself. If `sz isa StaticInt`, it'd be convenient to know the size, so we may want to pass it as an arg at some point (i.e., define another function to call) to get the parameters of the `OptionallyStaticUnitRange`.

Also, I'd always match a `pointer(src)` with a `GC.@preserve src`.

My plan is to have LoopVectorization automate the optimization on this stuff, so PaddedMatrices (for example) can use simpler optimized definitions using LoopVectorization.

Reply:
> Also, I'd always match a pointer(src) with a GC.@preserve src.

Good point

> My plan is to have LoopVectorization automate the optimization on this stuff, so PaddedMatrices (for example) can use simpler optimized definitions using LoopVectorization.

That sounds great! I'm just trying to tie together these different components of the interface into a cohesive product. Hopefully that will make it easier to get other packages integrated with this interface.

Perhaps a better approach would be something like:

```julia
function unsafe_get_collection(::CPUPointer, A, inds)
    axs = to_axes(A, inds)
    src = pointer(A)
    len = prod(map(static_length, inds))
    dst = similar_pointer(src, len)
    GC.@preserve A, unsafe_copyto_index!(dst, One():len, src, tdot(A, inds))
    return unsafe_reconstruct(A, dst; axes=axs)
end

@inline function unsafe_copyto_index!(dst, dst_itr, src, src_iter)
    dst_i = iterate(dst_itr)
    src_i = iterate(src_itr)
    while dst_i !== nothing
        unsafe_store!(dst, unsafe_load(src, first(src_i)), first(dst_i))
        dst_i = iterate(dst_itr, last(dst_i))
        src_i = iterate(src_itr, last(src_i))
    end
end

function similar_pointer(x, sz)
    return Base.unsafe_convert(Ptr{eltype(x)}, Libc.malloc(8 * sizeof(eltype(x)) * Int(sz)))
end
```

This uses `VectorizationBase.tdot` to tie this all together. It looks like there's a good amount of code that is necessary to support it, but it seems like the best way to provide support for array types in base that return `CPUPointer`. I'd also need to figure out how to support arrays that have non-bits type elements (eg., `Array{Any}`), but then LoopVectorization could use LLVM magic in its own `unsafe_copyto_index!` method for strided pointers.

Alternatively, we could push off supporting any `CPUPointer` indexing to the vectorization packages. I think there's a huge advantage to having something that ties these traits together right out of the gait. However, if this minor convenience makes it harder for you to maintain LoopVectorization and friends then I'm totally fine with focusing on generic `CPUPointer`  support elsewhere.

Reply:
`VectorizationBase.tdot` is much more complicated than you need here.
I'm sure with some work it'd be possible to simplify that implementation while keeping its functionality intact, but we don't need most of what it's meant to handle here.
That complexity stems from trying to make sure LLVM generates good pointer addressing code.
Basically, this is an AVX512 matmul kernel LoopVectorization generates:
```asm
L928:
        vmovupd zmm29, zmmword ptr [rbp]
        vmovupd zmm28, zmmword ptr [rbp + 64]
        vmovupd zmm27, zmmword ptr [rbp + 128]
        prefetcht0      byte ptr [rbp + r12 - 128]
        vbroadcastsd    zmm30, qword ptr [rcx - 8]
        prefetcht0      byte ptr [rbp + r12 - 64]
        prefetcht0      byte ptr [rbp + r12]
        vfmadd231pd     zmm0, zmm29, zmm30      # zmm0 = (zmm29 * zmm30) + zmm0
        vfmadd231pd     zmm1, zmm28, zmm30      # zmm1 = (zmm28 * zmm30) + zmm1
        vbroadcastsd    zmm31, qword ptr [rbx + rsi]
        vfmadd231pd     zmm2, zmm27, zmm30      # zmm2 = (zmm27 * zmm30) + zmm2
        vfmadd231pd     zmm3, zmm29, zmm31      # zmm3 = (zmm29 * zmm31) + zmm3
        vfmadd231pd     zmm4, zmm28, zmm31      # zmm4 = (zmm28 * zmm31) + zmm4
        vfmadd231pd     zmm5, zmm27, zmm31      # zmm5 = (zmm27 * zmm31) + zmm5
        vbroadcastsd    zmm30, qword ptr [rbx + 2*rsi]
        vfmadd231pd     zmm6, zmm29, zmm30      # zmm6 = (zmm29 * zmm30) + zmm6
        vfmadd231pd     zmm7, zmm28, zmm30      # zmm7 = (zmm28 * zmm30) + zmm7
        vfmadd231pd     zmm8, zmm27, zmm30      # zmm8 = (zmm27 * zmm30) + zmm8
        vbroadcastsd    zmm30, qword ptr [rbx + r15]
        vfmadd231pd     zmm9, zmm29, zmm30      # zmm9 = (zmm29 * zmm30) + zmm9
        vfmadd231pd     zmm10, zmm28, zmm30     # zmm10 = (zmm28 * zmm30) + zmm10
        vfmadd231pd     zmm11, zmm27, zmm30     # zmm11 = (zmm27 * zmm30) + zmm11
        vbroadcastsd    zmm30, qword ptr [rbx + 4*rsi]
        vfmadd231pd     zmm12, zmm29, zmm30     # zmm12 = (zmm29 * zmm30) + zmm12
        vfmadd231pd     zmm13, zmm28, zmm30     # zmm13 = (zmm28 * zmm30) + zmm13
        vfmadd231pd     zmm14, zmm27, zmm30     # zmm14 = (zmm27 * zmm30) + zmm14
        vbroadcastsd    zmm30, qword ptr [rbx + r14]
        vfmadd231pd     zmm15, zmm29, zmm30     # zmm15 = (zmm29 * zmm30) + zmm15
        vfmadd231pd     zmm16, zmm28, zmm30     # zmm16 = (zmm28 * zmm30) + zmm16
        vfmadd231pd     zmm17, zmm27, zmm30     # zmm17 = (zmm27 * zmm30) + zmm17
        vbroadcastsd    zmm30, qword ptr [rbx + 2*r15]
        vfmadd231pd     zmm18, zmm29, zmm30     # zmm18 = (zmm29 * zmm30) + zmm18
        vfmadd231pd     zmm19, zmm28, zmm30     # zmm19 = (zmm28 * zmm30) + zmm19
        vbroadcastsd    zmm31, qword ptr [rbx + rax]
        vfmadd231pd     zmm20, zmm27, zmm30     # zmm20 = (zmm27 * zmm30) + zmm20
        vfmadd231pd     zmm21, zmm29, zmm31     # zmm21 = (zmm29 * zmm31) + zmm21
        vfmadd231pd     zmm22, zmm28, zmm31     # zmm22 = (zmm28 * zmm31) + zmm22
        vfmadd231pd     zmm23, zmm27, zmm31     # zmm23 = (zmm27 * zmm31) + zmm23
        vbroadcastsd    zmm30, qword ptr [rcx + 8*rsi - 8]
        vfmadd231pd     zmm24, zmm30, zmm29     # zmm24 = (zmm30 * zmm29) + zmm24
        vfmadd231pd     zmm25, zmm30, zmm28     # zmm25 = (zmm30 * zmm28) + zmm25
        vfmadd231pd     zmm26, zmm27, zmm30     # zmm26 = (zmm27 * zmm30) + zmm26
        add     rbp, r13
        mov     rbx, rcx
        add     rcx, 8
        cmp     rbp, rdx
        jbe     L928
```
It is not perfect, but much better than it was half a year ago. Focusing on just the broadcasts:
```asm
        vbroadcastsd    zmm30, qword ptr [rcx - 8]
        vbroadcastsd    zmm31, qword ptr [rbx + rsi]
        vbroadcastsd    zmm30, qword ptr [rbx + 2*rsi]
        vbroadcastsd    zmm30, qword ptr [rbx + r15]
        vbroadcastsd    zmm30, qword ptr [rbx + 4*rsi]
        vbroadcastsd    zmm30, qword ptr [rbx + r14]
        vbroadcastsd    zmm30, qword ptr [rbx + 2*r15]
        vbroadcastsd    zmm31, qword ptr [rbx + rax]
        vbroadcastsd    zmm30, qword ptr [rcx + 8*rsi - 8]
```
These load 1 scalar from the memory location in brackets, e.g. from the pointer `rcx` minus `8` (`[rcx - 8]`), and copy it across a SIMD register.
The memory addressing lets us do a fairly complicated expression:
`a * x + c + y`, where `a` and `c` have to be constants (literal integers), while `x` and `y` can be integers/pointers. `a` is only allowed to be `0`, `1`, `2`, `4`, or, `8`.
These loads are across strides of an array, i.e.
`B[i,j]`, `B[i,j+1]`, `B[i,j+2]`, `B[i,j+3]`, ...

So basically, `c` is a pointer to `B[i,j]`, and then `x` is the stride in bytes.
This allows us to load from `B[i,j+1]` (a = 1), `B[i,j+2]` (a=2), `B[i,j+4]` (a = 4), and `B[i,j+8]`.
Then I have to also precalculate values equal to 3x the stride, 5x the stride, and 7x the stride, getting
`B[i,j+3]` (3x stride, a = 1), `B[i,j+6]` (3x stride, a = 2), then the +5 and +7 with `a = 1`.

LLVM will not automatically do anything like this, so it needs to be heavy handed in how the code gen coerces it.
Also, the code is imperfect, because for some reason there's both `rcx` and `rbx`, where `rcx == rbx + 8`, so it subtracts `8` every time it uses `rcx` instead of `rbx`. I have no idea why the compiler would think that's a good idea, but it means we end up with
```asm
        mov     rbx, rcx
        add     rcx, 8
```
when all we should have is
```asm
        add     rbx, 8
```
and replace all the `rcx - 8`s with `rbx`. But until I figure out how to convince LLVM to do that, I'll live with it.

Also, something else it has to do is avoid promoting the `MM` type (which corresponds to vector loads/stores) into a `Vec` type (which corresponds to slow gather/scatters), unless that is what the indexing requires.

But all this stuff, which the complexity comes from, isn't really necessary here.

> I think there's a huge advantage to having something that ties these traits together right out of the gait.

I agree. I don't think people should have to depend on `LoopVectorization` for support (my goal is just to make people want to ;) ).
I think it'd be good to provide an efficient default, but make it possible for people to opt into an alternative method.